### PR TITLE
fix: require pip upgrade to at least a recent version

### LINKED ISSUE #256: CI flake: pip update may not do anything
Probably should add a test forcing the exact minimum supported pip. The pip changelog is terrible about not having the details of when things get added or removed (try to find when eggs were dropped, for example).

—-

I believe this requires Pip 19.something (0 or 1), we need to bump the minimum check. It’s currently breaking our CI (macOS Python 3.6 has 18.something which is missing this flag).

_Originally posted by @henryiii in https://github.com/pypa/build/issues/232#issuecomment-794832786_

### PR DESCRIPTION
**Pull request by henryiii**
This will force the pip upgrade to update to a fairly recent version; if there's a network issue and PyPI can't be reached, this should provide a much clearer error message than simply having something later on break due to old versions. Closes #256.

### CODE REVIEW

**Review by FFY00 — COMMENTED** [review]
Hum, we should probably have a proper test for this. Shouldn't be that hard, just create an isolated env and look at the pip version.

Also, please change the commit prefix to `env:` :stuck_out_tongue:

**Comment by layday** [comment]
It doesn't print anything to stdout with the version constraint for me:

```
❯ python -m pip install -U pip
Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10b403280>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10b403ca0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10b3ece20>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10b3ec5b0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10b3ec5e0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
Requirement already up-to-date: pip in ./venv/lib/python3.9/site-packages (18.1)

❯ python -m pip install -U pip>=20.3.4
Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10c942250>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10c942d60>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10c92bd30>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10c92b550>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10c92b640>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
```

pytest didn't capture the retry messages from stderr either in https://github.com/pypa/build/runs/2068846663?check_suite_focus=true - are you sure connectivity was the issue?

**Comment by layday** [comment]
`-I` seems to have the desired effect, instead of `-U`:

```
❯ python -m pip install -I pip
Collecting pip
  Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10a71a040>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
  Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10a6e6df0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
  Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10a6cf5e0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
  Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10a6cfbe0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
  Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x10a71a310>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /simple/pip/
  Could not find a version that satisfies the requirement pip (from versions: )

❯ echo $status
1
```

**Comment by layday** [comment]
ahem, `-U 'pip>=20.3.4'` worked when I remembered to quote it but `-I` might still be the better choice here, rather than having to remember to track the version number in two different places?

**Comment by henryiii** [comment]
Let me ask a potentially idiotic "let's step back and look at this" question. Why not replace the whole check with `pip install pip>=21.0.1` (or whatever we require) in the first place, where we check currently? I just tried it - it doesn't require a network connection at all if the dependency is already met, and upgrades it if it is not, with a nice error if there's no network. That way we could kill off all our special pip version checks, too. Is there something special I'm missing?

**Comment by henryiii** [comment]
Pushed a comment to show what I'm thinking.

**Comment by layday** [comment]
The main reason would be not to call pip if not necessary which takes at least 200 ms to import in a new process. Try timing the two approaches.

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by henryiii** [comment]
The absolute minimum time for a basic package is around 5-6 seconds for an sdist or wheel, so the 0.2-0.3s is not very measurable. But maybe a hybrid approach would work well, let me try that.

**Review by layday — APPROVED** [review]
That looks good to me.  I'll leave it to others to decide what tests might be needed.

**Comment by FFY00** [comment]
To speed up things we could use `importlib.metadata` on Python 3.8 to find the pip version first, but that can come in a future PR as it is not really related to the change.

File: src/build/env.py [hunk 1]
```diff
@@ -219,24 +219,19 @@ def _create_isolated_env_venv(path):  # type: (str) -> Tuple[str, str]
...
 
-            if int(platform.mac_ver()[0].split('.')[0]) >= 11:
-                needs_pip_upgrade = pip_version < packaging.version.Version('20.3') or (
-                    pip_version < packaging.version.Version('21.0.1')  # Apple Silicon macOS requires 21.0.1+
-                    and sys.version_info >= (3, 6)
-                    and platform.machine() != 'x86_64'
-                )
-        if needs_pip_upgrade:
-            subprocess.check_call([executable, '-m', 'pip', 'install', '-U', 'pip'])
+        if current_pip_version < packaging.version.Version(pip_version):
```
Comment (layday) [inline src/build/env.py:233]: `minimum_pip_version` iso `pip_version`

**Comment by henryiii** [comment]
> To speed up things we could use importlib.metadata on Python 3.8 to find the pip version first, but that can come in a future PR as it is not really related to the change.

That's what the latest version of this does. I first replaced the version checking with just a simple pip call, which would add .2s to venv + new enough embedded pip, and then I added back the metadata version check to the restructured versions, which kept the simpler structure, but also should be as fast as it was before.

File: src/build/env.py (hunk 1 above)
Comment (henryiii) [reply src/build/env.py]: Done, also merged history so this is ready whenever everyone is happy with it.

**Comment by henryiii** [comment]
Are we waiting on anything here? @gaborbernat and @FFY00 is this good to go?

**Review by gaborbernat — APPROVED** [review]
